pkg/service/utgen/coverage.go

```diff
@@ -0,0 +1,138 @@
+package utgen
+
+import (
+	"encoding/xml"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+)
+
+// CoverageProcessor handles the processing of coverage reports
+type CoverageProcessor struct {
+	FilePath     string
+	Filename     string
+	CoverageType string
+}
+
+// NewCoverageProcessor initializes a CoverageProcessor object
+func NewCoverageProcessor(filePath, filename, coverageType string) *CoverageProcessor {
+	return &CoverageProcessor{
+		FilePath:     filePath,
+		Filename:     filename,
+		CoverageType: coverageType,
+	}
+}
+
+// ProcessCoverageReport verifies the report and parses it based on its type
+func (cp *CoverageProcessor) ProcessCoverageReport(timeOfTestCommand int64) ([]int, []int, float64, []string, error) {
+	err := cp.VerifyReportUpdate(timeOfTestCommand)
+	if err != nil {
+		return nil, nil, 0, nil, err
+	}
+	return cp.ParseCoverageReport()
+}
+
+// VerifyReportUpdate verifies the coverage report's existence and update time
+func (cp *CoverageProcessor) VerifyReportUpdate(timeOfTestCommand int64) error {
+	if _, err := os.Stat(cp.FilePath); os.IsNotExist(err) {
+		return fmt.Errorf("fatal: coverage report \"%s\" was not generated", cp.FilePath)
+	}
+
+	fileInfo, err := os.Stat(cp.FilePath)
+	if err != nil {
+		return err
+	}
+	fileModTimeMs := fileInfo.ModTime().UnixNano() / int64(time.Millisecond)
+
+	if fileModTimeMs <= timeOfTestCommand {
+		return fmt.Errorf("fatal: the coverage report file was not updated after the test command. file_mod_time_ms: %d, time_of_test_command: %d", fileModTimeMs, timeOfTestCommand)
+	}
+	return nil
+}
+
+// ParseCoverageReport parses the coverage report based on its type
+func (cp *CoverageProcessor) ParseCoverageReport() ([]int, []int, float64, []string, error) {
+	switch cp.CoverageType {
+	case "cobertura":
+		return cp.ParseCoverageReportCobertura()
+	case "lcov":
+		return nil, nil, 0, nil, fmt.Errorf("parsing for %s coverage reports is not implemented yet", cp.CoverageType)
+	default:
+		return nil, nil, 0, nil, fmt.Errorf("unsupported coverage report type: %s", cp.CoverageType)
+	}
+}
+
+type Coverage struct {
+	XMLName  xml.Name  `xml:"coverage"`
+	Sources  []string  `xml:"sources>source"`
+	Packages []Package `xml:"packages>package"`
+}
+
+type Package struct {
+	Name    string  `xml:"name,attr"`
+	Classes []Class `xml:"classes>class"`
+}
+
+type Class struct {
+	Name     string `xml:"name,attr"`
+	FileName string `xml:"filename,attr"`
+	Lines    []Line `xml:"lines>line"`
+}
+
+type Line struct {
+	Number int `xml:"number,attr"`
+	Hits   int `xml:"hits,attr"`
+}
+
+func (cp *CoverageProcessor) ParseCoverageReportCobertura() ([]int, []int, float64, []string, error) {
```

Comment:
This method returns too many things. Please make this an object

Reply:
refactored